### PR TITLE
Use Docker parser when manipulating Dockerfiles

### PR DESCRIPTION
**Pull request by kargakis**
Add one more layer of parsing for safety by using the Docker
parser to verify the valid occurrence of a Dockerfile command.

### CODE REVIEW

File: pkg/build/builder/docker.go [hunk 1]
```diff
@@ -171,6 +176,112 @@ func (d *DockerBuilder) addBuildParameters(dir string) error {
...
+	var replaceNextLn bool
+	for {
+		line, err := buf.ReadString('\n')
+		if err != nil && err != io.EOF {
+			return "", err
+		}
+		line = strings.TrimSpace(line)
+
+		// The current line starts with the specified command (cmd)
+		if strings.HasPrefix(line, cmd) || strings.HasPrefix(line, strings.ToUpper(cmd)) {
```
Comment (bparees) [inline pkg/build/builder/docker.go]: why not just strings.HasPrefix(strings.ToUpper(line),strings.ToUpper(cmd)) ?

**Comment by bparees** [comment]
No testcase for dockerfiles with trailing backslashes?

**Comment by bparees** [comment]
regarding the thread about validating the output, i'm not sure it's necessary given that the next thing we're going to do with the file is "docker build" it so if it's invalid, we'll get the error there...  us checking it beforehand doesn't seem to buy us much.... @soltysh ?

**Comment by kargakis** [comment]
> No testcase for dockerfiles with trailing backslashes?

Thats' why I have the `TODO` label at the thread post ;)

File: pkg/build/builder/docker.go (hunk 1 above)
Comment (kargakis) [inline pkg/build/builder/docker.go]: Cool, updating it...

**Comment by kargakis** [comment]
@bparees please review the latest changes and if you are okay with them, I'll squash the pr into two separate commits and rebase into current master

**Comment by bparees** [comment]
lgtm.  @smarterclayton any concern w/ us pulling in the upstream docker change here?

**Comment by smarterclayton** [comment]
Let me look

**Comment by ncdc** [comment]
Why is https://github.com/kargakis/origin/commit/e4048b996421f9d87aefc58da5eddfdb8648351d marked UPSTREAM?

**Comment by bparees** [comment]
@ncdc because it's pulling in upstream docker changes?

**Comment by ncdc** [comment]
My understanding is we UPSTREAM for things we want in our vendored copy of Kubernetes but that either a) aren't actually committed upstream yet, or b) are committed upstream, but we haven't rebased them in yet. See https://github.com/openshift/origin/blob/master/HACKING.md#3-cherry-pick-upstream-changes-pushed-to-the-origin-repo

**Comment by bparees** [comment]
this is the latter case, just that upstream in this case is docker not k8s.

**Comment by ncdc** [comment]
The point I'm trying to make is we aren't rebasing Docker, so having a commit with UPSTREAM that's not related to Kubernetes might lead to confusion when someone does a Kubernetes rebase.

**Comment by bparees** [comment]
got it.  but it seems like we have a similar problem that needs to be solved for docker rebases.  maybe a slightly different keyword is in order.

**Comment by ncdc** [comment]
Sure. But if all we're doing is running `godep update <package>`, we don't need UPSTREAM. In fact, we need a bump commit.

**Comment by kargakis** [comment]
IMHO marking a commit as UPSTREAM or bump or whatever is nothing compared to the merge commits cluttering the history logs right now but that's another story

**Comment by bparees** [comment]
@kargakis it's all about what is greppable.

**Comment by smarterclayton** [comment]
Yeah, part of our rebase process is double checking the bumps, so grepability is critical.

**Comment by soltysh** [comment]
> regarding the thread about validating the output, i'm not sure it's necessary given that the next thing we're going to do with the file is "docker build" it so if it's invalid, we'll get the error there... us checking it beforehand doesn't seem to buy us much.... @soltysh ?

My rationale here was as follows. Think of a use case where user provides us with valid Dockerfile but after modification we spit one that will not run build correctly, because of Dockerfile validation. This gets then reported back to user and he can't do nothing with it other than filling a bug. What I'm targeting here is we need to ensure that whatever comes out of our replace function must be valid Dockerfile implying the input was also valid. If we're tightening here the parsing let's go one step further and make sure that whatever comes out is 100% valid or we're getting replacement error, IMHO.

**Comment by ncdc** [comment]
@kargakis would you mind modifying the UPSTREAM commit to say bump instead?

**Comment by kargakis** [comment]
> @kargakis would you mind modifying the UPSTREAM commit to say bump instead?

No problem, done

**Comment by ncdc** [comment]
@kargakis thanks. I did just notice that the Docker commit you're pulling in isn't from a tag. @smarterclayton do you want the Docker commit to come from a tag, or is this ok?

**Comment by bparees** [comment]
@soltysh they can't do anything about it either way is my point, and it's going to get caught either way.  but sure, catching the error a little sooner where we can control the reported error message is not a bad idea.

**Comment by soltysh** [comment]
@bparees since we allowed those parameters then it should produce workable output, otherwise we failed at the parameter check step, which is other way of seeing this things.

**Comment by kargakis** [comment]
> What I'm targeting here is we need to ensure that whatever comes out of our replace function must be valid Dockerfile implying the input was also valid.

Aren't the existing tests accomplishing that?

**Comment by bparees** [comment]
@kargakis what @soltysh is suggesting isn't a new test case, it's output validation.  If someone provides an image name like "my image" then the final Dockerfile would contain "FROM my image" which presumably is invalid.

(this means we should probably have more validation on what is valid for the image field, if we don't already, but the point is we can end up with an invalid Dockerfile in a variety of ways and @soltysh thinks we should check for that rather than punt to docker build to catch the error)

**Comment by kargakis** [comment]
> @kargakis what @soltysh is suggesting isn't a new test case, it's output validation. If someone provides an image name like "my image" then the final Dockerfile would contain "FROM my image" which presumably is invalid.
> 
> (this means we should probably have more validation on what is valid for the image field, if we don't already, but the point is we can end up with an invalid Dockerfile in a variety of ways and @soltysh thinks we should check for that rather than punt to docker build to catch the error)

Ok, got it. Isn't it kind of irrelevant to my changes though? The functionality implemented here replaces Dockerfiles commands, while validating the image field is another story.

**Comment by kargakis** [comment]
Bitten by merge conflicts but resolved them yesterday. Also rebased on top of the current master.

**Comment by bparees** [comment]
it's not about validating just the image field, it's about validating the whole file to confirm we haven't accidentally corrupted it during our modification process (eg suppose there's some bug in your replacement logic).  so after doing the replace, you should confirm the new Dockerfile is valid.

**Comment by kargakis** [comment]
> it's not about validating just the image field, it's about validating the whole file to confirm we haven't accidentally corrupted it during our modification process (eg suppose there's some bug in your replacement logic). so after doing the replace, you should confirm the new Dockerfile is valid.

Ok, I am going to add that test I had in mind ie. parsing the old and new Dockerfiles and compare their ASTs. I will also add tests on `traverseAST`.

**Comment by kargakis** [comment]
Just added tests for comparing the Abstract Syntax Trees of the original and the edited Dockerfiles. Later I will write tests for `traverseAST`.

**Pull request closed without merging**